### PR TITLE
kernel: upgrade guest kernel support to 5.19

### LINKED ISSUE #4860: Upgrade guest kernel support to 5.19
**Which feature do you think can be improved?**
Guest kernel used can be upgraded to newly released 5.19.

**How can it be improved?**
Upgrading the guest kernel provides fixes for known issues and support for new features.

### PR DESCRIPTION
**Pull request by ryansavino**
Fixes #4860

kernel: upgrade guest kernel support to 5.19

Signed-Off-By: Ryan Savino <ryan.savino@amd.com>

### CODE REVIEW

**Comment by ryansavino** [comment]
This is blocked by #4849. That change is needed to build the 5.19 .config without any issues.

File: tools/packaging/kernel/build-kernel.sh
Comment (fidencio) [inline tools/packaging/kernel/build-kernel.sh:147]: I'm not sure if I understand this part here.  Why we would like to **not** get the kernel version from the `tee` entry in case of SEV?

File: tools/packaging/kernel/configs/fragments/x86_64/mmu.conf
Comment (fidencio) [inline tools/packaging/kernel/configs/fragments/x86_64/mmu.conf:4]: You're already setting this as part part of the base.conf

File: tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf
Comment (fidencio) [inline tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf:7]: Please, keep this sorted alphabetically.

File: versions.yaml
Comment (fidencio) [inline versions.yaml:156]: I'd prefer one commit bumping the kernel version, and another commit bumping the version for SEV, it'd be cleaner which change is related to each one of the targets.

**Review by fidencio — CHANGES_REQUESTED** [review]
I've left some commets, @ryansavino.

File: tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf
Comment (jongwu) [reply tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf:7]: Just out of curiosity. If kata always use direct kernel boot, what does efi stub kernel use for?

File: tools/packaging/kernel/build-kernel.sh
Comment (ryansavino) [reply tools/packaging/kernel/build-kernel.sh:147]: SEV can use the same base 5.19 kernel as general kata as all changes required for the specific kernel fork have been upstreamed. `get_tee_kernel` has some specifics regarding retrieving the version and associating it with the name of the tarball and the link to retrieve it. As kernel for SEV is the same as general kata now, I thought it best to have the script treat it the same. As an alternative, I can add handling in `get_tee_kernel` to fix the download link and the tarball name for SEV specifically.

File: tools/packaging/kernel/configs/fragments/x86_64/mmu.conf
Comment (ryansavino) [reply tools/packaging/kernel/configs/fragments/x86_64/mmu.conf:4]: Errors were being thrown during the .config file generation due to dependency config options not being present:
`CONFIG_PAGE_TABLE_ISOLATION=y` in mmu.conf has a dependency on `CONFIG_SPECULATION_MITIGATIONS=y`.
`CONFIG_RETPOLINE=y` in base.conf has a dependency on `CONFIG_SPECULATION_MITIGATIONS=y`.

I thought it best to add the same dependency in both respective files. This would avoid any .config file generation issues if one of these files was not included in the .config generation for some reason. It has no effect on the end .config as all options are merged in and redundancies are removed.

File: tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf
Comment (ryansavino) [reply tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf]: https://www.kernelconfig.io/CONFIG_EFI_STUB?q=CONFIG_EFI_STUB&kernelversion=5.19

This kernel feature allows a bzImage to be loaded directly
by EFI firmware without the use of a bootloader.

Perhaps it should be added to a different conf?

Comment (ryansavino) [reply tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf]: @fidencio, pushed the alphabetization.

File: versions.yaml
Comment (ryansavino) [reply versions.yaml:156]: done

**Comment by ryansavino** [comment]
It's worth mentioning here that the 5.15.x patches have several arm64 changes. Was wondering if someone from that side can help port these patches to the 5.19.x folder? @jongwu, got any bandwidth to help out with this? I'm suspecting certain CI tests will probably fail otherwise for arm64 testing.

File: tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf
Comment (jongwu) [reply tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf]: AFAIK, there is no kernel inside kata image. Am I right? @fidencio . So, where does firmware load the EFI stub kernel?

**Comment by jongwu** [comment]
> It's worth mentioning here that the 5.15.x patches have several arm64 changes. Was wondering if someone from that side can help port these patches to the 5.19.x folder? @jongwu, got any bandwidth to help out with this? I'm suspecting certain CI tests will probably fail for arm64 testing.

Ok, I will take a look at this.

**Comment by jongwu** [comment]
@ryansavino -, the patches under 5.15/arm-experimental are for experimental use on Arm. It will not impact the main kernel upgrade.

File: tools/packaging/kernel/build-kernel.sh
Comment (fidencio) [reply tools/packaging/kernel/build-kernel.sh:147]: If SEV can be used from the upstream kernel, I'd recommend removing the SEV entry from the versions.yaml file.  The only reason it was added was because the upstream kernel didn't have all the needed patches.

Comment (fidencio) [reply tools/packaging/kernel/build-kernel.sh:147]: Hmm. Thinking a little bit more about it I'm not sure if we should have SEV support enabled by default on the guest kernel, thus having the separate path would still make sense.  I need to think a little bit more about this.

File: tools/packaging/kernel/configs/fragments/common/dax.conf
Comment (fidencio) [inline tools/packaging/kernel/configs/fragments/common/dax.conf:15]: Why are you removing this one?

File: tools/packaging/kernel/configs/fragments/whitelist.conf
Comment (fidencio) [inline tools/packaging/kernel/configs/fragments/whitelist.conf:17]: This change is not needed / shouldn't be part of the PR.

File: tools/packaging/kernel/kata_config_version
Comment (fidencio) [inline tools/packaging/kernel/kata_config_version:1]: 95, but I'd do the bump in the next commit.

**Review by fidencio — CHANGES_REQUESTED** [review]
For the "kernel: upgrade guest kernel support to 5.19", I'd suggest the following commit message:
```
kernel: Upgrade guest kernel support to 5.19

Let's update to the latest 5.19.x released kernel, and while doing this,
let's also enable the `CONFIG_SPECULATION_MITIGATIONS`, which is
requiired to enable `RETPOLINE and `PAGE_TABLE_ISOLATION`.

Fixes #4860
```

File: tools/packaging/kernel/build-kernel.sh
Comment (fidencio) [reply tools/packaging/kernel/build-kernel.sh:147]: So, no.  The kernel is the same, but the options for `SEV` are only being applied in case of `conf_guest` is set to `sev`, and I think we should keep it that way for the time being.

So, please, remove this part of the commit.

File: tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf
Comment (fidencio) [inline tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf:2]: Not related to the PR.

**Review by fidencio — CHANGES_REQUESTED** [review]
I'd do the version bump as part of the "kernel: SEV guest kernel upgrade to 5.19" commit.

And I suggest the following commit message:
```
kernel: Update SEV guest kernel to 5.19.1

Kernel 5.19.1 has all the needed patches for running SEV, thus let's update it
and stop using the version coming from confidential-containers.
```

Also, I think this one also deserves an issue as that will help folks tracking the confidential-containers effort to easily pinpoint this change when needed.

File: versions.yaml
Comment (fidencio) [reply versions.yaml:156]: Also, please, update to v5.19.1

File: versions.yaml
Comment (fidencio) [inline versions.yaml:164]: Also update this one to v5.19.1

File: tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf
Comment (dubek) [reply tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf]: My understanding is:

In direct boot (with `-kernel`), QEMU passes the content of the kernel (and initrd, and cmdline) to the guest via the fw_cfg interface. OVMF (firmware) reads the kernel from that interface, places it in memory, and jumps to the kernel's EFI stub.

Since SEV (and TDX) must have OVMF (or TDVF), the kernels for these guests must have `CONFIG_EFI_STUB=y`.

File: versions.yaml
Comment (dubek) [reply versions.yaml:164]: I think it should be `version:` instead of `tag:`, because you're downloading from the CDN (and not from github).

File: tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf
Comment (dubek) [inline tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf:7]: I understand the CRYPTO settings are for the host and not needed in the guest; good.

What about CONFIG_SECURITYFS? Our efi_secret module (`CONFIG_EFI_SECRET=m`) uses it. Is it enabled in some other place?

File: tools/packaging/kernel/kata_config_version
Comment (ryansavino) [reply tools/packaging/kernel/kata_config_version]: done

File: versions.yaml
Comment (ryansavino) [reply versions.yaml:156]: done

File: tools/packaging/kernel/configs/fragments/common/dax.conf
Comment (ryansavino) [reply tools/packaging/kernel/configs/fragments/common/dax.conf:15]: CONFIG_DEV_PAGEMAP_OPS is no longer configurable: https://www.kernelconfig.io/config_dev_pagemap_ops

Will add to commit message.

File: tools/packaging/kernel/configs/fragments/whitelist.conf
Comment (ryansavino) [reply tools/packaging/kernel/configs/fragments/whitelist.conf:17]: Added back the newline. Done.

File: tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf
Comment (ryansavino) [reply tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf:2]: Re-added the space on the line.

Comment (ryansavino) [reply tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf:2]: Also, updated the commit message.

File: versions.yaml
Comment (ryansavino) [reply versions.yaml:164]: Version updated. @dubek you mind if we address the version/tag item in a separate issue?

File: tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf
Comment (ryansavino) [reply tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf:7]: @dubek I was attempting to cleanup the sev.conf by removing all unnecessary CONFIG options. I confirmed that the list I currently provide in this PR is the minimal set, and I successfully booted an SEV kata container using a 5.19 guest. I'm using the command `./build-kernel.sh -x sev setup` to generate the .config. CONFIG_SECURITYFS is making it's way into the final .config. I'm not sure if something else is adding it, or it's there by default now? I'm happy to undo it's removal in this PR if you feel it necessary.

File: tools/packaging/kernel/build-kernel.sh
Comment (ryansavino) [reply tools/packaging/kernel/build-kernel.sh]: I've removed this `conf_guest` check for `sev` and added a check in `get_tee_kernel` to see if `kernel_url` contains the string `cdn.kernel.org`. If it finds the generic kernel URL, it prepends `linux-` at the front or the tarball name. Tried to do this in the most generic way possible. Let me know if you'd like me to modify it differently.

File: tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf
Comment (ryansavino) [reply tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf:7]:
```
root@rysavino-ubuntu-02:~/kata-containers/tools/packaging/kernel/configs# grep -R "CONFIG_SECURITYFS" .
./arm64_kata_kvm_4.14.x:# CONFIG_SECURITYFS is not set
./arm64_kata_kvm_virtio-fs-v0.3.x:# CONFIG_SECURITYFS is not set
./powerpc_kata_kvm_4.19.x:# CONFIG_SECURITYFS is not set
./arm64_kata_kvm_4.19.x:# CONFIG_SECURITYFS is not set
./s390_kata_kvm_5.4.x:CONFIG_SECURITYFS=y
./fragments/x86_64/.config:CONFIG_SECURITYFS=y
./arm64_kata_kvm_5.4.x:# CONFIG_SECURITYFS is not set
./powerpc_kata_kvm_5.4.x:# CONFIG_SECURITYFS is not set
./x86_64_kata_kvm_4.14.x:# CONFIG_SECURITYFS is not set
./s390_kata_kvm_4.19.x:# CONFIG_SECURITYFS is not set
./powerpc_kata_kvm_4.14.x:# CONFIG_SECURITYFS is not set
```

Maybe default option? From above search it's making its way into fragments/x86_64/.config, but isn't being set in any conf that would affect x86_64.

Comment (dubek) [reply tools/packaging/kernel/configs/fragments/x86_64/sev/sev.conf:7]: If it's in the final `.config` I think it's fine for now (maybe CONFIG_EFI_SECRET selects it, I need to look it up).

**Comment by ryansavino** [comment]
`kata deploy build / make-kata-tarball (pull_request)` is failing due to #4786 and #4787

File: tools/packaging/kernel/configs/fragments/x86_64/mmu.conf
Comment (fidencio) [reply tools/packaging/kernel/configs/fragments/x86_64/mmu.conf:4]: Please. drop this one from here.
The `base` config will **always** be used, not making much sense to have this one duplicated here.

Comment (fidencio) [reply tools/packaging/kernel/configs/fragments/x86_64/mmu.conf:4]: And also from the commit message.

**Comment by fidencio** [comment]
@ryansavino, just a minor comment about dropping the duplicated config, as it's not needed.

Please, update that, including the commit message, and this one will be ready to be merged.

**Review by fidencio — APPROVED** [review]
lgtm, thanks @ryansavino!

**Comment by fidencio** [comment]
cc @wainersm for a look on this as well.

**Comment by fidencio** [comment]
@ryansavino, the kernel upgrade is surely causing some issues with the CI.  Do you mind debugging it?

**Comment by dubek** [comment]
Hmmm, I found this error in the log of jenkins-ci-ubuntu-20.04 :

```
Running nydus integration tests
Image is up to date for sha256:dcac4621b8636ac822ab7737648a272fcf378b8df6d315816036f00d5536d1e2
Pod 6bc88318657fef6d6855c02dd1fdf5a2a094731ce210e406059ad9d3f7cd4ce8 created
Container d44191eddc024a568c3c944a8c2d57f18308a8550a6a30b51389441710749041 created
time="2022-08-18T07:21:50Z" level=fatal msg="starting the container \"d44191eddc024a568c3c944a8c2d57f18308a8550a6a30b51389441710749041\": rpc error: code = Unknown desc = failed to create containerd task: failed to create shim: unable to write to a control group file caused by: Os { code: 95, kind: Uncategorized, message: \"Not supported\" }: unknown"
```

Maybe related to #4390 (affects linux >= 5.16)? I already submitted the fix and it was merged in a new cgroups-rs release, but kata-agent wasn't updated yet.

**Comment by ryansavino** [comment]
> @ryansavino, the kernel upgrade is surely causing some issues with the CI. Do you mind debugging it?

Sure, let me look.

**Comment by ryansavino** [comment]
Confirmed that I can create test-lifecycle-events.yaml with no errors after upgrading agent cgroup-rs crate to 0.2.10. Added commit and pushed. Thanks for the tip @dubek.

**Comment by fidencio** [comment]
@BbolroC, any chance to help us to understand the failures in the s390x CI?

**Comment by wainersm** [comment]
> @BbolroC, any chance to help us to understand the failures in the s390x CI?

This PR deleted [tools/packaging/kernel/configs/fragments/s390/network.conf](https://github.com/kata-containers/kata-containers/pull/4861/commits/32960d340905b57cf08c16e1be866e1a2f8589bf#diff-3d519fef6459ab438b480a7d6a049ee2b25a252a9d1f1cdcd5d03dcbc9066af4)

By mistake @ryansavino ?

**Comment by wainersm** [comment]
> > @BbolroC, any chance to help us to understand the failures in the s390x CI?
> 
> This PR deleted [tools/packaging/kernel/configs/fragments/s390/network.conf](https://github.com/kata-containers/kata-containers/pull/4861/commits/32960d340905b57cf08c16e1be866e1a2f8589bf#diff-3d519fef6459ab438b480a7d6a049ee2b25a252a9d1f1cdcd5d03dcbc9066af4)
> 
> By mistake @ryansavino ?

It seems `CONFIG_PACK_STACK ` is still supported by s390. Where did you find it is not?

File: tools/packaging/kernel/build-kernel.sh
Comment (wainersm) [reply tools/packaging/kernel/build-kernel.sh]: @ryansavino I've a suggestion to address the tarball name issue. I will comment in line.

File: tools/packaging/kernel/build-kernel.sh
Comment (wainersm) [inline tools/packaging/kernel/build-kernel.sh:130]: My suggestion is to assume `kernel_tarball="linux-${version}.tar.gz"` is expected, otherwise add the 'tarball' property in the `versions.yml`. Something like:

```
kernel_tarball="linux-${version}.tar.gz"
tarball_name=$(get_from_kata_deps "assets.kernel.${tee}.tarball")
[ -z "$tarball_name" ] || kernel_tarball="$tarball_name"
```
Then change the TDX properties in `versions.yaml`:
```
diff --git a/versions.yaml b/versions.yaml
index eb6753293..b76f6e104 100644
--- a/versions.yaml
+++ b/versions.yaml
@@ -158,6 +158,7 @@ assets:
       description: "Linux kernel that supports TDX"
       url: "https://github.com/intel/linux-kernel-dcp/archive/refs/tags"
       tag: "SPR-BKC-PC-v9.6"
+      tarball: "SPR-BKC-PC-v9.6.tar.gz"
     sev:
       description: "Linux kernel with efi_secret support"
       url: "https://github.com/confidential-containers-demo/\
```

File: tools/packaging/kernel/build-kernel.sh
Comment (wainersm) [inline tools/packaging/kernel/build-kernel.sh:562]: Avoid the use of `A && B` because that expression can return non-zero status. Instead use `! A || B` which always return zero (ah, unless `B` return fails.. which is not the case here).

**Comment by wainersm** [comment]
@ryansavino I don't have more to say than what @fidencio and @dubek already said.

**Comment by BbolroC** [comment]
@fidencio Hi, unfortunately due to the internal situation, I am currently supporting other projects urgently. I could only fix any broken CI jobs caused by infra problems for the time being (maybe for a month at least). Hopefully I could come back as fast as possible.

**Comment by fidencio** [comment]
> > > @BbolroC, any chance to help us to understand the failures in the s390x CI?
> > 
> > 
> > This PR deleted [tools/packaging/kernel/configs/fragments/s390/network.conf](https://github.com/kata-containers/kata-containers/pull/4861/commits/32960d340905b57cf08c16e1be866e1a2f8589bf#diff-3d519fef6459ab438b480a7d6a049ee2b25a252a9d1f1cdcd5d03dcbc9066af4)
> > By mistake @ryansavino ?
> 
> It seems `CONFIG_PACK_STACK ` is still supported by s390. Where did you find it is not?

Nice catch, I missed this.

File: tools/packaging/kernel/build-kernel.sh
Comment (ryansavino) [reply tools/packaging/kernel/build-kernel.sh:130]: done

File: tools/packaging/kernel/build-kernel.sh
Comment (ryansavino) [reply tools/packaging/kernel/build-kernel.sh]: done

**Comment by wainersm** [comment]
> > > > @BbolroC, any chance to help us to understand the failures in the s390x CI?
> > > 
> > > 
> > > This PR deleted [tools/packaging/kernel/configs/fragments/s390/network.conf](https://github.com/kata-containers/kata-containers/pull/4861/commits/32960d340905b57cf08c16e1be866e1a2f8589bf#diff-3d519fef6459ab438b480a7d6a049ee2b25a252a9d1f1cdcd5d03dcbc9066af4)
> > > By mistake @ryansavino ?
> > 
> > 
> > It seems `CONFIG_PACK_STACK ` is still supported by s390. Where did you find it is not?
> 
> Nice catch, I missed this.

I was chatting with @ryansavino on private about it. On https://www.kernelconfig.io/CONFIG_PACK_STACK?q=CONFIG_PACK_STACK&kernelversion=5.19.2 it is not clear whether that config is still supported or not. If that config is include then the `.config` is not generated as you can see in http://jenkins.katacontainers.io/job/kata-containers-2.0-ubuntu-s390x-main-PR/2210/consoleFull

Omitting that config, it builds the kernel but fails to copy the `vmlinux` file at  
https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/kernel/build-kernel.sh#L458 . That script expects the vmlinux on `arch/s390/boot/compressed/vmlinux`...

**Comment by ryansavino** [comment]
I'll need access to an s390 system in order to debug further. Looks like the `jenkins-ci-s390x-unit-ubuntu-20-04` job is failing because it can't find the C++ compiler. The `jenkins-ci-s390x-main-ubuntu-20-04` and the `jenkins-ci-s390x-snap-ubuntu-20-04` jobs are failing because 'arch/s390/boot/compressed/vmlinux' isn't present. @wainersm any way I can get access to the s390 CI node directly?

**Comment by BbolroC** [comment]
@fidencio @wainersm I am supporting Ryan atm for debugging. Hopefully, we could make the best out of it. :wink:

**Comment by BbolroC** [comment]
@ryansavino https://github.com/kata-containers/kata-containers/blob/main/snap/snapcraft.yaml#L209 should be changed as well?

**Comment by ryansavino** [comment]
Opened a tests issue (https://github.com/kata-containers/tests/issues/5047) regarding failing `jenkins-ci-s390x-unit-ubuntu-20-04`.

**Comment by fidencio** [comment]
Once this one lands, we will need to raise a backport for the 5.19 bump, the SEV work should **not** be backported though.

**Review by wainersm — APPROVED** [review]

**Comment by wainersm** [comment]
> Once this one lands, we will need to raise a backport for the 5.19 bump, the SEV work should **not** be backported though.

hi @fidencio , I didn´t get why we need to backport the kernel update. Is it fixing bugs?

**Comment by fidencio** [comment]
> > Once this one lands, we will need to raise a backport for the 5.19 bump, the SEV work should **not** be backported though.
> 
> hi @fidencio , I didn´t get why we need to backport the kernel update. Is it fixing bugs?

It brings in the `CONFIG_SPECULATION_MITIGATIONS`, which would be welcome in the stable branch.

**Comment by dubek** [comment]
> It brings in the `CONFIG_SPECULATION_MITIGATIONS`, which would be welcome in the stable branch.

I think we can update the stable branch to latest in 5.15.y -- it should have the RETbleed mitigations (5.15 is the most recent LTS kernel). (but 5.15 doesn't have efi_secret kernel module which is needed for SEV support)


On another subject: @fidencio  How do we get the changes from this PR in the CCv0 branch? Do we regularly rebase CCv0 on top of main?

**Comment by dubek** [comment]
I see that `main` was merged into branch `CCv0`, so we now have 5.19.2 in CCv0 as well. (Thanks @Megan-Wright )

**Pull request merged**
